openedxscorm/scormxblock.py

```diff
@@ -336,24 +337,38 @@ def extract_package(self, package_file):
                 raise ScormError(
                     "Could not find 'imsmanifest.xml' file in the scorm package"
                 )
-
-            for zipinfo in zipinfos:
-                # Extract only files that are below the root
-                if zipinfo.filename.startswith(root_path):
-                    # Do not unzip folders, only files. In Python 3.6 we will have access to
-                    # the is_dir() method to verify whether a ZipInfo object points to a
-                    # directory.
-                    # https://docs.python.org/3.6/library/zipfile.html#zipfile.ZipInfo.is_dir
-                    if not zipinfo.filename.endswith("/"):
-                        dest_path = os.path.join(
-                            self.extract_folder_path,
-                            os.path.relpath(zipinfo.filename, root_path),
-                        )
-                        self.storage.save(
-                            dest_path,
-                            ContentFile(scorm_zipfile.read(zipinfo.filename)),
+            else:
+                self.zip_root_path = root_path
+        
+            max_workers = self.xblock_settings.get("THREADPOOLEXECUTOR_MAX_WORKERS", 10)
```

Comment:
Is unzipping each file one by one the part that is causing very slow response times?

I'd like to avoid resorting to multithreading, if possible. It's a little too complex for my taste. Could we unzip locally to a temporary directory before uploading the folder in one go?

Reply:
Yes that issue reporter mentioned slow upload speed over S3. While uploading scorm package we extract it and then each file one by one. It appears extracting(unzipping) the files is not taking that much time because reporter mentioned speed on local storage is fine. These changes introduce parallel upload to S3 to reduce upload time.

Reply:
Did you manage to reproduce the issue? I understand it would be a bit tricky to setup an S3 bucket just for that purpose.

Reply:
Hi and thanks for mentioning me!

I'll try the fix ASAP (most probably on the next week)

Reply:
@dyudyunov have you been able to test it?

Reply:
Hello @ziafazal 

Sorry for the delay, I've got a lot of stuff going on

I can verify now that I can successfully upload a 5.9MB file using your fix! 🎉
Thank you!

Reply:
upload time was close to 30sec according to logs

Reply:
Yay! Now I'm wondering whether we can find a solution that does not involve multithreading. Is there a native storage API that would support copying an entire directory, instead of copying data file by file?

Reply:
It appears that in django-storages>=1.14 which is being used in Quince release it is using [Threading](https://github.com/boto/boto3/blob/bae6340b70a8fb1e96debcc21fa37a14a41d2805/boto3/s3/transfer.py#L164) to optimize upload of files on S3.

Reply:
So, do we still have to implement threading ourselves or is it handled natively by django-storages?

Reply:
As discuss plan is to implement it ourselves for pre quince version and In quince we should remove our implementation and rely on multi threading support of django-storages.